posts/2022-04-15-libertydevops-gitjenkinshelmiks.adoc

```diff
@@ -0,0 +1,195 @@
+
+* Maven, to build java code
+* Pipeline, for creating Jenkins pipeline jobs
+* Multibranch Scan Webhook Trigger, to create Jenkins pipeline type jobs which pull all branches of code from Git to Jenkins
+* Docker, to build and push container images
+* Kubernetes, to use the Kubernetes template
+
+Jenkins builds the Liberty Java code using Jenkins pipeline scripts. The script can run directly on your Jenkins installation (known as Jenkins master) but, if the application is big, you need to use a Jenkins agent. For more information about about setting up Jenkins agents, see link:https://www.jenkins.io/doc/book/using/using-agents/[Using Jenkins agents] and link:[https://www.jenkins.io/doc/book/pipeline/syntax/Pipeline Syntax].
+
+== Writing Jenkins pipeline scripts
+
+It is good practice to adhere to the concept of infrastructure as a code (IaaC) when writing DevOps pipelines. Writing Jenkins jobs as pipeline scripts is a good example of IaaC.
+
+You can write link:https://www.jenkins.io/doc/book/pipeline/syntax/[Jenkins pipeline code] in one of the following ways:
+
+* As pipeline code written directly in the link:https://www.jenkins.io/doc/book/pipeline/getting-started/[Jenkins UI] and stored in the Jenkins master. This is a useful way to get started. You would need to take a backup of the Jenkins instance to store the pipeline code because it is part of the Jenkins installation. 
+
+[.imageblock.img_border_light]
+image::/img/blog/pipeline-code-on-jenkins.png[Pipeline code directly on Jenkins,width=70%,align="center"]
+
+* As plain text in a Jenkinsfile (a plain text file) in Git and mapping it to Jenkins. This is better for ensuring that your configuration is always under version control. If you have specific build and deployment configurations for separate environments, such as dev, staging, and production, you create a separate Jenkinsfile for each environment and store it in that environment-specific Git repository branches. You can use either the "Pipeline" or "Multibranch pipeline" type of job in this case.
+
+[.imageblock.img_border_light]
+image::/img/blog/pipeline-code-on-git.png[Pipeline code stored in a Jenkinsfile on git,width=50%,align="center"]
+
+== Building the cloud-native Java application with Open Liberty on Jenkins
+
+Build your Dockerfile from the foundation code given below. Additional files mentioned in the code can be downloaded from link:/extras/featureUtility.properties[featureUtility.properties] , link:/extras/server.xml[server.xml] 
+
+[source]
+----
+FROM icr.io/appcafe/open-liberty:full-java11-openj9-ubi
+# Add Liberty server configuration including all necessary features
+COPY --chown=1001:0 server.xml /config/
+# Modify feature repository (optional), this file contains the proxy settings to connect to private maven repository
+COPY --chown=1001:0 featureUtility.properties /opt/ol/wlp/etc/
```

Comment:
If you think it's important to keep these lines in, could we add a sentence after the script saying

"If you are using a private Maven repository, see link:https://openliberty.io/docs/latest/reference/command/featureUtility-commands.html[featureUtility commands]."

Reply:
Alternatively, just remove these lines from the script (and still provide the link to the docs).